### ISSUE TITLE
parsePacket(buffer) not giving expected output.

### ISSUE THREAD

**Issue by jerbear2008** [SkeldJS/SkeldJS#7]
I am trying to use [`cap`](https://www.npmjs.com/package/cap) and `@skeldjs/protocol` to decode player list and player position. I have a constant stream of buffers that I am sending through `parsePacket()` and logging the response to try to figure out how it works. 

My code and the result of running it were uploaded to a new Gist [here](https://gist.github.com/jerbear2008/607e9c0f2cc5d3a8c0faecffce48f8c8).
Looking at the [auto-updating docs](https://skeldjs.github.io/SkeldJS/enums/constant.opcode.html), which aren't very helpful, the `op` value should be 10, 9, 8, 12, 1, or 0, not the numbers I am getting.
I know Cap is working correctly because I only get output when connecting to the internet from Among Us and I get more when more is happening.

Am I using the wrong function to do this? Do I need to switch to TypeScript? I don't really understand how exporting interfaces to a vanilla javascript file works.

**Comment by jerbear2008** [comment]
For now, I think I will move to [among.js](https://github.com/kognise/among-js) because of it's simplicity, and because my only options are wait for help or switch.

**Comment by jerbear2008** [comment]
Among.js [isn't working](https://github.com/kognise/among-js/issues/15) right now, so @edqx if you could help me debug this, that would be great.

**Comment by edqx** [comment]
Hi, I'm really not sure what you mean, are you sure the `cap` data is giving you actual UDP payload? or is it perhaps giving you the entire packet?
Also if you could elaborate on the docs not being very helpful that would be great.

**Comment by edqx** [comment]
There's an example on the README of the package that you're using to get the UDP data. https://github.com/mscdex/cap#examples

You should also use `parsePacket` rather than `parsePayload` if you need to parse the entire thing.

**Comment by edqx** [comment]
Also considering `among-js`'s last commit was 7 months ago, it probably wouldn't parse your packets correctly anyway.

**Comment by jerbear2008** [comment]
I am using `Cap.open(device, filter, bufSize, buffer)`, which "Opens device and starts capturing packets using filter" and "buffer is a Buffer large enough to store one packet", and because there are examples that decode wether or not it is even UDP, I think it just captures the whole packet. Is that the problem? There is this comment in the first example, saying `// raw packet data === buffer.slice(0, nbytes)`. Should I pass that into parsePacket? Should I use code like:
```js
if (linkType === 'ETHERNET') {
  var ret = decoders.Ethernet(buffer);
  if (ret.info.type === PROTOCOL.ETHERNET.IPV4 && ret.info.protocol === PROTOCOL.IP.UDP) {
    console.log('Packet');
    ret = decoders.UDP(buffer, ret.offset);
    let data = parsePacket(buffer.slice(ret.offset, ret.offset + ret.info.length));
  }
}
```
Also, the docs don't make things very clear, like, what is an opcode? I can't find it online.

**Comment by jerbear2008** [comment]
> Also considering `among-js`'s last commit was 5 months ago, it probably wouldn't parse your packets correctly anyway.

Yeah, fully moved away from among-js.

**Comment by edqx** [comment]
I'm unsure how the package that you're using to record packets works or what the API looks like so I won't be able to help you much there, although the code that you sent would seem to me to work well.

An example of how the `@skeldjs/protocol` package can be used effectively can be seen in the client source here: https://github.com/SkeldJS/SkeldJS/blob/bc72408903fcacf2604f1b22e526941414facf48/packages/client/ts/lib/client.ts#L250

Keep in mind however, as you did mention that you hope to record the player list, that you will have to keep track of a lot of things yourself using the [JoinGame](https://github.com/codyphobe/among-us-protocol/blob/master/02_root_message_types/01_joingame.md) (for when players join) and [RemovePlayer](https://github.com/codyphobe/among-us-protocol/blob/master/02_root_message_types/04_removeplayer.md) (for when players leave), as well as recording player data like their name with [SetName](https://github.com/codyphobe/among-us-protocol/blob/master/04_rpc_message_types/06_setname.md) and their colour with [SetColour](https://github.com/codyphobe/among-us-protocol/blob/master/04_rpc_message_types/08_setcolor.md).
Additionally, recording player positions will require having to deserialize data especially for a specific object that you will have to keep track of spawning and despawning.

You can either implement all of it yourself, as the protocol package itself comes with no state management, or you can quite easily use the `@skeldjs/client` package to do that (without actually making the client connect to a server & join games), let me know if you need help with that.

In regards to the docs, I've linked several keywords to the repo at https://github.com/codyphobe/among-us-protocol, which describes the protocol and is helpful if  you want to learn more about the protocol. The docs for skeldjs are simply there to reflect the usage API, not to describe the protocol. (Note that some things are renamed)

**Comment by jerbear2008** [comment]
Wow, thanks! I have never seen a package maintainer so helpful! Yeah, using `@skeldjs/client` seems like a good idea, I'll try to figure that out.

**Comment by jerbear2008** [comment]
This seems really useful. I could run this whenever a UDP packet is captured on an Among Us port:
```js
let parsed = decoders.UDP(buffer, ret.offset);
data = parsePacket(buffer.slice(ret.offset, ret.offset + ret.info.length));
if (data.bound === "client") client.handlePacket(data)
console.info("packet ")
```
However, the [handlePacket](https://github.com/SkeldJS/SkeldJS/blob/bc72408903fcacf2604f1b22e526941414facf48/packages/client/ts/lib/client.ts#L257) function seems to ack all the packets it receives that require one. How do I make the code passively observe instead of responding?
#https://github.com/SkeldJS/SkeldJS/blob/bc72408903fcacf2604f1b22e526941414facf48/packages/client/ts/lib/client.ts#L257
Again, thanks for being so helpful!

**Comment by edqx** [comment]
This is actually a valid point, I should add a way to do this, and ideally I should remake my `@skeldjs/state` package that I had a while ago but removed in favour of using the client package instead.

There's 2 options for this, you can either re-create the `handlePacket` method but without handling the Acknowledge or Ping packets, and change the state externally or you can extend the `SkeldjsClient` class and override `send()` to do nothing. Either will work.

Thanks for bringing this to my attention, I'll refactor this later and see about bringing back the state package for convenience.

**Comment by edqx** [comment]
Also I forgot to mention in case you weren't aware - you can use the events that are propagated through all of the objects and systems to the client to listen for changes. There's a short list here: https://github.com/SkeldJS/SkeldJS/blob/master/spec/events.md but I haven't found a way to auto-document all of the events in any useful way.

**Comment by jerbear2008** [comment]
Oh, that's helpful. Is there any way I can disable the client sending packets in response to the server?

**Comment by edqx** [comment]
Not sure what you mean, you can override the `send()` method in `SkeldjsClient` by extending the class, or you can wait ~10 minutes until I publish a new `@skeldjs/state` package.

**Comment by jerbear2008** [comment]
I think I found a workaround. If the `client.send()` function finds that `client.socket` does not exist, it will `return null`. So, if I run `client.socket = false` before I call `client.handlePacket()`, it should not send.
https://github.com/SkeldJS/SkeldJS/blob/bc72408903fcacf2604f1b22e526941414facf48/packages/client/ts/lib/client.ts#L584
I'll try this.

**Comment by edqx** [comment]
There shouldn't be a socket anyway, surely? Unless you're calling `.connect()`. But that's not necessary if you just call `await .onMessage(` with the data from your captured packets. You don't need to connect to the server or join a game.

**Comment by jerbear2008** [comment]
Oh, that makes sense. I will test it out!

**Comment by jerbear2008** [comment]
My code so far:
```js
import * from "@skeldjs/protocol"
import * from "@skeldjs/client"
let client = new SkeldjsClient("2021.4.14s");

let Cap = require("cap").Cap;
let decoders = require("cap").decoders;
let PROTOCOL = decoders.PROTOCOL;

let c = new Cap();
let device = "\\Device\\NPF_{3964EFF5-5E70-442F-BD2A-9F7BF07E89E8}";
let filter = "udp portrange 22023-22923";
let bufSize = 10 * 1024 * 1024;
let buffer = Buffer.alloc(65535);
let linkType = c.open(device, filter, bufSize, buffer);
c.setMinBytes && c.setMinBytes(0);

c.on("packet", function(nbytes, trunc) {
  try {
    let parsed;
    if (!trunc) {
      let parsed = decoders.UDP(buffer, ret.offset);
      let data = parsePacket(buffer.slice(ret.offset, ret.offset + ret.info.length));
      if (data.bound === "client") {
        client.handlePacket(data)
        console.info("packet")
      }

    } else {
      console.log("!!!     truncated packet");
    }
  } catch (err) {
    console.error(err);
  }
});






const ev = new require("events")();
let something = 0;
ev.on("increment", () => {
  something++
});
const timer = setInterval(() => ev.emit("increment"), 1e5);
// keep node from exiting while waiting for packets
```

**Comment by edqx** [comment]
Pretty sure you don't even need to `parsePacket` either, you can just use call the `onMessage` function with the buffer.

Regarding the import syntax, you should probably use `require()` since you use that everywhere else.
```js
const { SkeldjsClient } = require("@skeldjs/client");
```
You don't need to import `@skeldjs/protocol` pretty sure.

**Comment by jerbear2008** [comment]
Well, I was decoding it with parsePacket because handlePacket requires a clientbound packet. Now that I think about it, I can update my Cap filter to only accept inbound packets.

**Comment by edqx** [comment]
The `bound` in `parsePacket` and the inbound/outbound status of your packets aren't related, so it wouldn't make a difference. `parsePacket` parses packet data as though it was sent to the client unless stated otherwise.

It has just hit me though, your program won't record some changes if you're the host and I can go into detail for that if you want, but basically if you only listen for packets from the server, some changes that happen locally and only get sent to the server and broadcasted to every other client won't affect the state of your program since it's not getting the full picture.

You might have to give the program both inbound and outbound packets in order for it to get the full picture. I might need to look into this further and try it myself.

**Comment by jerbear2008** [comment]
Ok, I will pass it all, and it will probably be fine because it has a try{} around it

**Comment by edqx** [comment]
Can never go wrong with a `try {} catch { ignore }`.

**Comment by jerbear2008** [comment]
Got an error:
```
D:\node\Among Us\node_modules\@skeldjs\core\js\lib\system\MovingPlatformSystem.js:48
        writer.uint8(this.target?.control?.netid ?? null);
                                 ^

SyntaxError: Unexpected token '.'
←[90m    at wrapSafe (internal/modules/cjs/loader.js:1053:16)←[39m
←[90m    at Module._compile (internal/modules/cjs/loader.js:1101:27)←[39m
←[90m    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1157:10)←[39m
←[90m    at Module.load (internal/modules/cjs/loader.js:985:32)←[39m
←[90m    at Function.Module._load (internal/modules/cjs/loader.js:878:14)←[39m
←[90m    at Module.require (internal/modules/cjs/loader.js:1025:19)←[39m
←[90m    at require (internal/modules/cjs/helpers.js:72:18)←[39m
    at Object.<anonymous> (D:\node\Among Us\node_modules\←[4m@skeldjs←[24m\core\js\lib\component\AirshipStatus.js:8:32)
←[90m    at Module._compile (internal/modules/cjs/loader.js:1137:30)←[39m
←[90m    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1157:10)←[39m
```
https://github.com/SkeldJS/SkeldJS/blob/bc72408903fcacf2604f1b22e526941414facf48/packages/core/ts/lib/system/MovingPlatformSystem.ts#L94

**Comment by edqx** [comment]
I think you just need to update your nodejs version if possible.

**Comment by jerbear2008** [comment]
Ok, give me a minute.

**Comment by jerbear2008** [comment]
Ok, it appears to work!

**Comment by edqx** [comment]
Wonderful, I'd still make sure it works entirely before you can say it works though. If everything works feel free to close the issue.

**Comment by jerbear2008** [comment]
There is a problem. All the values of client seem to be null or undefined.

**Comment by jerbear2008** [comment]
Ah, empty buffers are being passed to handlePacket.

**Comment by edqx** [comment]
I'm heading off for tonight, if you need anything I'll have a look in the morning.

**Comment by edqx** [comment]
Hi again, you will probably be interested to know that I recently published a [@skeldjs/state](https://github.com/SkeldJS/SkeldJS/tree/master/packages/state) [package](https://www.npmjs.com/package/@skeldjs/state) and will probably be more helpful than the client.

It can accept both an outbound and an inbound package with `handleInboundMessage` and `handleOutboundMessage` which you will need to decide between based on the direction of the packet that you capture.

I haven't really tested it 100% (just a small amount), but I can't really see where it would fail and I imagine it will work fine.

It has the same events that you can listen to as the client.

**Comment by jerbear2008** [comment]
Thanks, that sounds much better for my purposes.
I just used Wireshark to capture a clientbound test packet from among us:
`3c7c3fb1d3785c5f678273eb080045000032216e00008011d31dc0a83202800ed376c8e2585f001e1d3b00120005b858c48c0b0001050c00927e5485ff7fff7f`
The `00120005b858c48c0b0001050c00927e5485ff7fff7f` part is the actual Among Us payload that I need to pass to the code. It contains a game code ZPTFNF and player coordinates for the NetId 5. I will send this through instead of the actual packets to test.

**Comment by edqx** [comment]
You will also have to send every packet you receive up until that point.

**Comment by jerbear2008** [comment]
Ok, I was passing the wrong thing to the functions. This code works!
```js
const {
  SkeldjsStateManager
} = require("@skeldjs/state");
let state = new SkeldjsStateManager("2021.4.14s");

let Cap = require("cap").Cap;
let decoders = require("cap").decoders;
let PROTOCOL = decoders.PROTOCOL;
let udp = require('udp-packet')

let c = new Cap();
let device = Cap.findDevice('192.168.50.2'); // internal ip
let filter = "udp portrange 22023-22923";
let bufSize = 10 * 1024 * 1024;
let buffer = Buffer.alloc(65535);
let linkType = c.open(device, filter, bufSize, buffer);
c.setMinBytes && c.setMinBytes(0);

c.on("packet", function(nbytes, trunc) {
  try {
    if (!trunc) {
      let ret = decoders.Ethernet(buffer);
      ret.ipv4 = decoders.IPV4(buffer, ret.offset);
      ret.udp = decoders.UDP(buffer, ret.ipv4.offset);

      if (ret.ipv4.info.dstaddr.startsWith("192.168.")) {
        state.handleInboundMessage(buffer.subarray(ret.udp.offset, ret.udp.offset + ret.udp.info.length))
      } else {
        state.handleOutboundMessage(buffer.subarray(ret.udp.offset, ret.udp.offset + ret.udp.info.length))
      }
      console.log(state.players)
    }
  } catch (err) {
    console.error(err);
  }
});
```
I am so happy! I have been trying to get this to work for a week or so! Thanks for being so helpful!

**Comment by edqx** [comment]
This is actually useful code, I might clean it up and have it as an example somewhere. Glad to hear you got it working anyway, feel free to open another issue if something else goes wrong.

**Comment by jerbear2008** [comment]
The code was made in such a way that the local computer did not even have to be in the game, it can be another device on the network.
FYI, I'm planning to put my finished code on GitHub.

**Comment by edqx** [comment]
Actually that would be better if you did that, I could have a list of "projects made with skeldjs/example projects" and link yours.

**Comment by jerbear2008** [comment]
Ok, I will let you know when it is finished.

**Comment by jerbear2008** [comment]
The correct way to go about this seems to be to listen to the events and put the data into an object I can use in the rest of my code, but I cannot seem to find the correct way to listen to the events. Do I have to do something like:
```js
state.players.forEach((player, id, playerMap)=>{
  player.on("player.setname", setName)
})

function setName(name){
  if (!players[this.data.id]) players[this.data.id] = {}
  players[this.data.id].data = this.data
  console.log(name + " joined")
}
```
Are the events broadcasted from the [`SkeldjsStateManager`](https://skeldjs.github.io/SkeldJS/classes/state.skeldjsstatemanager.html) object or the individual [`PlayerData`](https://skeldjs.github.io/SkeldJS/classes/client.playerdata.html) objects? The docs either don't make this clear or I don't understand them.

**Comment by jerbear2008** [comment]
By the way, what I am coding is a version of Among Us proximity chat, but the players just connect to a website and talk through that. That way, even people without Among Us on their computers can play. Of course, one computer will have to be monitoring the game, using your code.

**Comment by edqx** [comment]
> The correct way to go about this seems to be to listen to the events and put the data into an object I can use in the rest of my code, 
You can just access the state for those things, you're just doing what the state is already doing.

> Are the events broadcasted from the SkeldjsStateManager object or the individual PlayerData objects?
They're broadcasted from the players but SkeldJS is designed to propagate everything "upwards", in the sense that you can listen for events from any player by listening to the state manager.

e.g.
```ts
state.on("player.setname", ev => {
    console.log("Player " + ev.data.player.id + " set their name to " + ev.data.name);
});
```
or you can listen to the player directly.
```ts
const player =state.players.get(1000);
player.on("player.setname", ev => {
    console.log("Player " + player.id + " set their name to " + ev.data.name);
});
```
notice how as it propagates upwards, it attaches information about where it came from, like `ev.data.player`

You're right that this should be more clear though.

**Comment by edqx** [comment]
So the code you wrote can be refactored to:
```ts
state.on("player.setname", ev => {
    console.log(ev.data.name + " joined the game.");
});
```

**Comment by edqx** [comment]
If you want to see an example of this "in the wild", there's loads of events here that you can see to get a better idea: https://github.com/auproximity/auproximity/blob/3e845d1160316281bcb3a8517ec2f652d2094c73/src/backends/PublicLobbyBackend.ts#L475

**Comment by jerbear2008** [comment]
Thanks for clarifying!

**Comment by edqx** [comment]
I think I just found a bug so the `player.setname` event won't work, I'll push a fix out.

**Comment by jerbear2008** [comment]
Ah, thanks. Also, I had no idea that AUProximity was a thing, and is pretty much the same thing I wanted to do, but so much better. I am still going to finish my project, because of the time it took.

**Comment by edqx** [comment]
You should still work on it, auproximity uses a bot to join the game rather than using cap. Cap wouldn't work for a program like auproximity where the main instance is public and where I can't verify the data received from users using cap.

**Comment by edqx** [comment]
Okay yea, fixed it, just released `@skeldjs/state 2.7.3`.
